File: ios/HaliSahaApp/HaliSahaApp/ViewModels/MapViewModel.swift
//
//  MapViewModel.swift
//  HaliSahaApp
//
//  Harita ViewModel - MapKit iş mantığı
//
//  Created by Mehmet Mert Mazıcı on 1.01.2026.
//

import Foundation
import SwiftUI
import MapKit
import Combine

// MARK: - Map ViewModel
@MainActor
final class MapViewModel: ObservableObject {
    
    // MARK: - Published Properties
    @Published var facilities: [Facility] = []
    @Published var facilitiesWithDistance: [(facility: Facility, distance: Double)] = []
    @Published var selectedFacility: Facility?
    @Published var isLoading = false
    @Published var error: String?
    
    // Map State
    @Published var region: MKCoordinateRegion
    @Published var mapCameraPosition: MapCameraPosition = .automatic
    
    // UI State
    @Published var showListView = false
    @Published var showFilters = false
    @Published var showFacilityDetail = false
    @Published var searchText = ""
    @Published var filters = FacilityFilters()
    
    // MARK: - Private Properties
    private let facilityService = FacilityService.shared
    private let locationManager = LocationManager.shared
    private var cancellables = Set<AnyCancellable>()
    
    // MARK: - Computed Properties
    var filteredFacilities: [Facility] {
        var result = facilities
        
        // Metin araması
        if !searchText.isEmpty {
            let query = searchText.lowercased()
            result = result.filter { facility in
                facility.name.lowercased().contains(query) ||
                facility.address.lowercased().contains(query)
            }
        }
        
        // Filtreler
        if filters.hasActiveFilters {
            if let isIndoor = filters.isIndoor {
                result = result.filter { $0.amenities.isIndoor == isIndoor }
            }
            if let minRating = filters.minRating {
                result = result.filter { $0.averageRating >= minRating }
            }
            if filters.hasParking {
                result = result.filter { $0.amenities.hasParking }
            }
            if filters.hasShower {
                result = result.filter { $0.amenities.hasShower }
            }
            if filters.hasCafe {
                result = result.filter { $0.amenities.hasCafe }
            }
        }
        
        return result
    }
    
    var annotations: [FacilityAnnotation] {
        filteredFacilities.map { FacilityAnnotation(facility: $0) }
    }
    
    var hasActiveFilters: Bool {
        !searchText.isEmpty || filters.hasActiveFilters
    }
    
    // MARK: - Init
    init() {
        // Varsayılan bölge (Ankara)
        self.region = MKCoordinateRegion(
            center: CLLocationCoordinate2D(
                latitude: AppConstants.defaultLatitude,
                longitude: AppConstants.defaultLongitude
            ),
            span: MKCoordinateSpan(
                latitudeDelta: 0.1,
                longitudeDelta: 0.1
            )
        )
        
        setupBindings()
    }
    
    // MARK: - Setup Bindings
    private func setupBindings() {
        // Konum değişikliklerini dinle
        locationManager.$userLocation
            .compactMap { $0 }
            .first() // Sadece ilk konumu al
            .sink { [weak self] location in
                self?.centerOnUserLocation(location)
            }
            .store(in: &cancellables)
    }
    
    // MARK: - Load Facilities
    func loadFacilities() async {
        isLoading = true
        error = nil
        
        do {
            // Mock data kullan (Firebase bağlantısı olmadığında)
            let loadedFacilities = facilityService.loadMockFacilities()
            
            // Gerçek implementasyon:
            // let loadedFacilities = try await facilityService.fetchAllFacilities()
            
            self.facilities = loadedFacilities
            
            // Mesafeleri hesapla
            calculateDistances()
            
            isLoading = false
        } catch {
            isLoading = false
            self.error = error.localizedDescription
        }
    }
    
    // MARK: - Calculate Distances
    private func calculateDistances() {
        let userLocation = locationManager.currentOrDefaultLocation
        
        facilitiesWithDistance = facilities.map { facility in
            let facilityLocation = CLLocationCoordinate2D(
                latitude: facility.latitude,
                longitude: facility.longitude
            )
            let distance = userLocation.distance(to: facilityLocation)
            return (facility, distance)
        }.sorted { $0.distance < $1.distance }
    }
    
    // MARK: - Request Location Permission
    func requestLocationPermission() {
        locationManager.requestPermission()
    }
    
    // MARK: - Center on User Location
    func centerOnUserLocation(_ location: CLLocationCoordinate2D? = nil) {
        let targetLocation = location ?? locationManager.userLocation ?? locationManager.defaultLocation
        
        withAnimation(.easeInOut(duration: 0.5)) {
            region = MKCoordinateRegion(
                center: targetLocation,
                span: MKCoordinateSpan(latitudeDelta: 0.05, longitudeDelta: 0.05)
            )
            mapCameraPosition = .region(region)
        }
        
        // Mesafeleri yeniden hesapla
        calculateDistances()
    }
    
    // MARK: - Center on Facility
    func centerOnFacility(_ facility: Facility) {
        withAnimation(.easeInOut(duration: 0.3)) {
            region = MKCoordinateRegion(
                center: facility.coordinate,
                span: MKCoordinateSpan(latitudeDelta: 0.02, longitudeDelta: 0.02)
            )
            mapCameraPosition = .region(region)
        }
    }
    
    // MARK: - Select Facility
    func selectFacility(_ facility: Facility?) {
        selectedFacility = facility
        
        if let facility = facility {
            centerOnFacility(facility)
            showFacilityDetail = true
        } else {
            showFacilityDetail = false
        }
    }
    
    // MARK: - Toggle View Mode
    func toggleViewMode() {
        withAnimation(.easeInOut(duration: 0.3)) {
            showListView.toggle()
        }
    }
    
    // MARK: - Clear Filters
    func clearFilters() {
        searchText = ""
        filters.reset()
    }
    
    // MARK: - Get Distance String
    func distanceString(for facility: Facility) -> String {
        if let match = facilitiesWithDistance.first(where: { $0.facility.id == facility.id }) {
            if match.distance < 1 {
                return String(format: "%.0f m", match.distance * 1000)
            } else {
                return String(format: "%.1f km", match.distance)
            }
        }
        return ""
    }
}

// MARK: - Facility Annotation
struct FacilityAnnotation: Identifiable {
    let id: String
    let facility: Facility
    var coordinate: CLLocationCoordinate2D
    
    init(facility: Facility) {
        self.id = facility.id ?? UUID().uuidString
        self.facility = facility
        self.coordinate = CLLocationCoordinate2D(
            latitude: facility.latitude,
            longitude: facility.longitude
        )
    }
}

// MARK: - Map Style
enum MapStyleOption: String, CaseIterable, Identifiable {
    case standard = "Standart"
    case satellite = "Uydu"
    case hybrid = "Hibrit"
    
    var id: String { rawValue }
    
    var mapStyle: MapStyle {
        switch self {
        case .standard: return .standard
        case .satellite: return .imagery
        case .hybrid: return .hybrid
        }
    }
}

File: android/app/src/main/java/com/example/HaliSahaApp/ui/viewmodels/MapViewModel.kt
package com.example.HaliSahaApp.ui.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.HaliSahaApp.data.models.Facility
import com.example.HaliSahaApp.data.models.UserLocation
import com.example.HaliSahaApp.data.services.FacilityFilters
import com.example.HaliSahaApp.data.services.FacilityService
import com.example.HaliSahaApp.data.services.LocationService
import com.example.HaliSahaApp.utils.AppConstants
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.LatLng
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

// MARK: - Map UI State
data class MapUiState(
    val facilities: List<Facility> = emptyList(),
    val filteredFacilities: List<Facility> = emptyList(),
    val selectedFacility: Facility? = null,
    val isLoading: Boolean = false,
    val error: String? = null,

    // UI Controls
    val showListView: Boolean = false,
    val showFilters: Boolean = false, // BottomSheet kontrolü için
    val searchText: String = "",
    val filters: FacilityFilters = FacilityFilters(),

    // Map State
    val userLocation: UserLocation? = null,
    val cameraPosition: CameraPosition? = null // Harita kamerasını güncellemek için
) {
    val hasActiveFilters: Boolean
        get() = searchText.isNotEmpty() || filters.hasActiveFilters
}

// MARK: - Map ViewModel
class MapViewModel : ViewModel() {

    private val facilityService = FacilityService
    private val locationService = LocationService

    private val _uiState = MutableStateFlow(MapUiState())
    val uiState: StateFlow<MapUiState> = _uiState.asStateFlow()

    init {
        // Varsayılan kamera pozisyonu (İstanbul)
        val defaultLatLng = LatLng(AppConstants.DEFAULT_LATITUDE, AppConstants.DEFAULT_LONGITUDE)
        _uiState.update {
            it.copy(cameraPosition = CameraPosition.fromLatLngZoom(defaultLatLng, AppConstants.DEFAULT_MAP_ZOOM))
        }

        setupBindings()

        viewModelScope.launch {
            loadFacilities()
        }
    }

    // MARK: - Setup Bindings (Konum Dinleme)
    private fun setupBindings() {
        viewModelScope.launch {
            locationService.userLocation.collectLatest { location ->
                _uiState.update { it.copy(userLocation = location) }

                // Eğer ilk kez konum geldiyse oraya odaklan
                if (location != null && _uiState.value.facilities.isEmpty()) {
                    centerOnUserLocation()
                }

                // Mesafeleri yeniden hesapla veya listeyi güncelle (opsiyonel)
                applyFilters()
            }
        }
    }

    // MARK: - Load Facilities
    private suspend fun loadFacilities() {
        _uiState.update { it.copy(isLoading = true, error = null) }

        try {
            // Şimdilik Mock Data (Service'deki)
            val loadedFacilities = facilityService.loadMockFacilities()
            // İleride: facilityService.fetchAllFacilities()

            _uiState.update {
                it.copy(
                    facilities = loadedFacilities,
                    isLoading = false
                )
            }
            applyFilters() // İlk yüklemede filtreleri uygula

        } catch (e: Exception) {
            _uiState.update { it.copy(isLoading = false, error = e.localizedMessage) }
        }
    }

    // MARK: - Filtering Logic
    private fun applyFilters() {
        val currentState = _uiState.value
        var result = currentState.facilities

        // Metin Araması
        if (currentState.searchText.isNotEmpty()) {
            val query = currentState.searchText.lowercase()
            result = result.filter { facility ->
                facility.name.lowercase().contains(query) ||
                        facility.address.lowercase().contains(query)
            }
        }

        // Filtreler
        val filters = currentState.filters
        if (filters.hasActiveFilters) {
            filters.isIndoor?.let { isIndoor ->
                result = result.filter { it.amenities.isIndoor == isIndoor }
            }
            filters.minRating?.let { minRating ->
                result = result.filter { it.averageRating >= minRating }
            }
            if (filters.hasParking) result = result.filter { it.amenities.hasParking }
            if (filters.hasShower) result = result.filter { it.amenities.hasShower }
            if (filters.hasCafe) result = result.filter { it.amenities.hasCafe }
        }

        // Sıralama (Opsiyonel: Yakına göre sırala)
        currentState.userLocation?.let { userLoc ->
            result = result.sortedBy { facility ->
                val facLoc = UserLocation(facility.latitude, facility.longitude)
                userLoc.distanceTo(facLoc)
            }
        }

        _uiState.update { it.copy(filteredFacilities = result) }
    }

    // MARK: - Actions

    fun onSearchTextChange(text: String) {
        _uiState.update { it.copy(searchText = text) }
        applyFilters()
    }

    fun toggleViewMode() {
        _uiState.update { it.copy(showListView = !it.showListView) }
    }

    fun toggleFilters() {
        _uiState.update { it.copy(showFilters = !it.showFilters) }
    }

    fun selectFacility(facility: Facility?) {
        _uiState.update { it.copy(selectedFacility = facility) }

        if (facility != null) {
            centerOnFacility(facility)
        }
    }

    fun clearFilters() {
        _uiState.update {
            it.copy(searchText = "", filters = FacilityFilters())
        }
        applyFilters()
    }

    // Filtre güncelleme (BottomSheet'ten gelecek)
    fun updateFilters(newFilters: FacilityFilters) {
        _uiState.update { it.copy(filters = newFilters) }
        applyFilters()
    }

    // MARK: - Map Actions

    fun centerOnUserLocation() {
        val location = _uiState.value.userLocation ?: return
        val target = LatLng(location.latitude, location.longitude)

        _uiState.update {
            it.copy(cameraPosition = CameraPosition.fromLatLngZoom(target, 14f))
        }
    }

    private fun centerOnFacility(facility: Facility) {
        val target = LatLng(facility.latitude, facility.longitude)
        _uiState.update {
            it.copy(cameraPosition = CameraPosition.fromLatLngZoom(target, 16f))
        }
    }

    fun requestLocationPermission(context: android.content.Context) {
        // İzin isteme mantığı UI (Screen) tarafında LaunchedEffect ile tetiklenmeli
        // veya Accompanist Permissions kütüphanesi kullanılmalı.
        // ViewModel sadece "izin istendi" state'ini tutabilir veya servise yönlendirebilir.
        locationService.startUpdatingLocation(context)
    }

    // MARK: - Helpers

    fun getDistanceString(facility: Facility): String {
        val userLoc = _uiState.value.userLocation ?: return ""
        val facLoc = UserLocation(facility.latitude, facility.longitude)
        val distance = userLoc.distanceTo(facLoc) // km döner

        return if (distance < 1.0) {
            String.format("%.0f m", distance * 1000)
        } else {
            String.format("%.1f km", distance)
        }
    }
}